### ISSUE TITLE
Layers instances should not contain any information about current render state

### ISSUE THREAD

**Issue by Czaki** [napari/napari#4682]
## 🚀 Feature

All data required only for rendering data (view on data subset) should be extracted to separate classes for simplifying layer reuse.

## Motivation

During the development of #4665 I found that create multiple view on same data require connection to multiple events and adds multiple obsolete calculations when synchronize attributes.
Life will be much simpler if Layers have zero knowledge about current view and then could be added to multiple viewers. 

## Pitch

all properties and methods like `_ImageBase.data_level`, `_ImageBase.plane`, `Plane._view_data` would be extracted to separate class that will take Layer class as constructor argument.  
Then data description could be used in multiple viewers without need for manual curation of synchronized things. 

## Alternatives

Plugin creators may manually trace changes in napari and do it on their own hand.

**Comment by alisterburt** [comment]
If I understand correctly the way we've been thinking about this is more that the layer data should be separated out and the layer should be the thing which keeps track of everything required to render a specific view of the data. Is that consistent with your view of the path forwards @andy-sweet ?

**Comment by Czaki** [comment]
No. But maybe also Data should be also put in the container. 
As you could see in #4665 I synchronize all things like scale, colormap etc. But still, layers have properties like currently selected slice, etc. 
And these things I would like to extract.

**Comment by jni** [comment]
Yup, slicing info needs to come out of the layer, we've been discussing this for some time and @andy-sweet's prototypes with asynchronous slicing are moving in this direction. But currently the private attributes are *very* enmeshed in the code which makes it a big undertaking.

**Comment by andy-sweet** [comment]
I roughly think of the state in a layer as follows.

1. The main substance of a layer that is important for processing and visualization: `data` (the VIP), `features` (if present), and `_transforms`.
2. The visualization style of a layer that tells us how to draw it in the canvas: `Layer.opacity`, `Layer.blending`, `Image.colormap`, `Points.face_color`, etc.
3. The currently selected elements (e.g. points, labels): `Points_selected_data`, maybe `Labels._selected_label`
4. The current view/slice of data: `Image._slice`, `Points.__indices_view`, probably `Image._plane`

I would expect (1), (2), and probably (3) to be equal across different views/slices of the same layer, at least in the case of ortho-view multi-canvas. For the recently discussed [spaces](https://github.com/napari/napari/pull/4684), I'm not so sure - if a layer is allowed to belong to more than one space then maybe it should have a space-specific transform and an independent selection. 

One example of the enmeshed state is `Points._selected_view` which is effectively an intersection of `Points.__indices_view` and `Points._selected_data` (i.e. 3 and 4). Though I'm sure there are other more complicated ones out there...

**Comment by brisvag** [comment]
> For the recently discussed [spaces](https://github.com/napari/napari/pull/4684), I'm not so sure - if a layer is allowed to belong to more than one space then maybe it should have a space-specific transform and an independent selection.

I agree with your separation of layer state, and I think it a good indication that *no*, transform cannot change between spaces. Transform *is* part of the data. As for selection, that's a bit trickier, and I could imagine a world where that lives on the `LayerSlice`.

**Comment by Czaki** [comment]
Transform is part of data. Selection is part of data visualization.
